syntax = "proto3";

package inbs;

service INBSSBService {
  // Bi-directional streaming method
  rpc INBMCommand(stream INBMResponse) returns (stream INBMRequest);
}

message INBMRequest {
  string request_id = 1;
  INBMRequestPayload payload = 2;
}

message INBMResponse {  
  string request_id = 1;
  INBMResponsePayload payload = 2;
}

message INBMRequestPayload {
  oneof payload {
    PingRequestPayload ping_request = 1;
    ScheduledUpdateRequestPayload scheduled_update_request = 2;
  }
}

message INBMResponsePayload {
  oneof payload {
    PingResponsePayload ping_response = 1;
    ScheduledUpdateResponsePayload scheduled_update_response = 2;
  }
}

message PingRequestPayload {
}

message PingResponsePayload {
}

message ScheduledUpdateRequestPayload {
  enum UpdateType {
    SOTA = 0; // Software Over The Air updates
  }
  UpdateType update_type = 1;
  oneof schedule {
    int64 epoch_time = 2; // Scheduled time in UNIX epoch seconds
    bool immediate = 3; // For immediate updates
  }
}

message ScheduledUpdateResponsePayload {
}